src/fsharp/FSharp.Core/seq.fs

```diff
@@ -842,6 +812,397 @@ namespace Microsoft.FSharp.Collections
     [<RequireQualifiedAccess>]
     [<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
     module Seq =
+//        type ISeqEnumerable<'T> =
+//            abstract member Fold<'State> : folder:('State->'T->'State) -> state:'State -> 'State
+
+        module SeqComposer =
+            open IEnumerator
+
+            module Helpers =
+                // used for performance reasons; these are not recursive calls, so should be safe
+                let inline avoidTailCall x =
+                    match x with
+                    | true -> true
+                    | false -> false
+
+                let inline ComposeFilter f g x = f x && g x
+
+            type [<AbstractClass>] SeqComponent<'T,'U> () =
+                abstract ProcessNext : input:'T * halt:byref<bool> * output:byref<'U> -> bool
+                abstract OnComplete : unit -> unit
+
+                abstract Composer : SeqComponent<'U,'V> -> SeqComponent<'T,'V>
+
+                abstract ComposeMap<'S>       : Map<'S,'T>       -> SeqComponent<'S,'U>
```

Comment:
Tricky naming; they way the functions work out is that it is ComposeMapWith as map is "first". That might be better... Maybe...